Bidirectional planner

pevnak opened the pull request:
I have implemented a simple bidirectional planner. The implementation is basic and uses most functionality from backward and forward planner. I have tried to preserve the style as much as possible. Unit tests added. I am very happy to improve the implementation as you wish.

Best wishes,
Tomas

ztangent commented:
Thanks so much for making this PR! It'll be awesome to have a bidirectional planner in this library, so thanks for getting started on it, and also maintaining the code style.

There are a couple of changes I'm going to suggest, but I think one more of the important ones is that the termination condition right now is subtly off. This isn't obvious from my implementation, but the `BackwardPlanner` actually operates over a different search space than the `ForwardPlanner`, because nodes in the `BackwardPlanner` search tree actually correspond to *partial* states (equivalent to sets of complete states), and so you can't check whether the search frontiers cross just by checking whether one of the forward search nodes ended up in the backward search tree (and vice versa).

Addressing this issue efficiently is non-trivial -- this paper (["Analyzing the Impact of Partial States on Duplicate Detection and Collision of Frontiers"](https://ojs.aaai.org/index.php/ICAPS/article/view/13677)) discusses it in some detail.

The simple (but inefficient) thing to do would be to check whether each state expanded by the forward search is a subset of every search node visited by the backward search visited so far, by using `issubset(forward_state, backward_state)`. This would require iterating over the entire backward search tree each time a node gets expanded, and so I think it would lose a lot of efficiency gains.

Another simple approach to try instead would be to check whether any newly expanded forward node is subsumed by one of the search nodes of the *open* set of the backward search (and vice versa). This would be a less expensive check, but wouldn't always be able to detect a frontier crossing.

All in all I'm leaning towards the second approach, just because it would be sometimes useful without being overly complicated to implement. Alternatively, if you want to implement some of the more complicated approaches described in the paper I linked, you're welcome to do that!

ztangent commented on src/planners/bidirectional.jl:
```diff
@@ -0,0 +1,130 @@
+export BiDirectionalPlanner, BiGreedyPlanner, BiAStarPlanner
+
+"BiDirectionalPlanner best-first search planner."
+mutable struct BiDirectionalPlanner{F,B} <: Planner
```
I think I would prefer this to be renamed `BidirectionalPlanner`, just because that's a bit more standard English-wise, though I recognize this means it would be less consistent with `BiGreedyPlanner` and `BiAStarPlanner`.

pevnak replied:
Done

ztangent commented on src/planners/bidirectional.jl:
```diff
@@ -0,0 +1,130 @@
+    BiDirectionalPlanner(
+    	ForwardPlanner(;heuristic=heuristic, kwargs...),
+    	BackwardPlanner(;heuristic=heuristic, kwargs...)
+    )
```
For both of these constructors, I think the forward heuristic and backward heuristic should actually be specified separately, rather than shared across both planners. This is because some heuristics, like `HAdd`, only make sense when used with a `ForwardPlanner`, whereas `HAddR` is only meant to be used with a `BackwardPlanner`.

Maybe replace with.

```julia
BiAStarPlanner(f_heuristic::Heuristic, b_heuristic::Heuristic; kwargs...) =
    BiDirectionalPlanner(
    	ForwardPlanner(;heuristic=f_heuristic, kwargs...),
    	BackwardPlanner(;heuristic=b_heuristic, kwargs...)
    )
```

I would also adjust all of the other constructors similarly.

ztangent commented on src/planners/bidirectional.jl:
```diff
@@ -0,0 +1,130 @@
+        backward_plan, backward_traj = reconstruct(node_id, backward_search_tree)
+        reverse!(backward_plan); reverse!(backward_traj)
+        plan = vcat(forward_plan, backward_plan)
+        traj = vcat(forward_traj, backward_traj)
```
So as mentioned in [this comment](https://github.com/JuliaPlanners/SymbolicPlanners.jl/pull/8#issuecomment-1345398608), the states explored by forward and backward search are actually different. This means you can't just concatenate the trajectories found by forward and backward search to get a single trajectory of complete states.

Instead, I think it's easier to just avoid storing the full trajectory as part of the `BiPathSearchSolution`. Alternatively, you could use the `StateTrajectorySimulator` to simulate the full trajectory, once you have the full plan.

ztangent commented on src/planners/bidirectional.jl:
```diff
@@ -0,0 +1,130 @@
+                 forward_spec::Specification, forward_search_tree::Dict{UInt,<:PathNode}, forward_queue::PriorityQueue,
+                 backward_spec::Specification, backward_search_tree::Dict{UInt,<:PathNode}, backward_queue::PriorityQueue)
+	max_nodes = max(planner.forward.max_nodes)
+	max_time = max(planner.forward.max_time)
```
Rather than using the forward planners parameters, I think maybe it'd be a good idea to store this separately in the `BiDirectionalPlanner` struct? Also, I'm a little confused by why you're using the `max` function here.

ztangent commented on src/planners/bidirectional.jl:
```diff
@@ -0,0 +1,130 @@
+        # Get state with lowest estimated cost to goal
+        node_id = dequeue!(forward_queue)
+        node = forward_search_tree[node_id]
+        if is_goal(forward_spec, domain, node.state) || node_id ∈ keys(backward_search_tree)
```
As mentioned [earlier](https://github.com/JuliaPlanners/SymbolicPlanners.jl/pull/8#issuecomment-1345398608), this ` node_id ∈ keys(backward_search_tree)` check isn't quite right, and will have to be adjusted to handle the fact that backward search states are partial states, not complete states.

ztangent commented on src/planners/bidirectional.jl:
```diff
@@ -0,0 +1,130 @@
+        # Get state with lowest estimated cost to goal
+        node_id = dequeue!(backward_queue)
+        node = backward_search_tree[node_id]
+        if is_goal(backward_spec, domain, node.state) || node_id ∈ keys(forward_search_tree)
```
See above comment.

Also,  I think it might be a good idea to check for frontier crossings only once per search iteration. So after expanding forward search, check if the goal is reached. After expanding backward search, check if the start state is reached. And only if both those checks fail, then check if the search frontiers have crossed.

ztangent commented on src/planners/path_search.jl:
```diff
@@ -47,3 +47,26 @@ Base.iterate(sol::PathSearchSolution) = iterate(sol.plan)
+mutable struct BiPathSearchSolution{S<:State,T} <: OrderedSolution
+    status::Symbol
+    plan::Vector{Term}
+    trajectory::Union{Vector{S},Nothing}
```
Maybe remove `trajectory` from the search solution, as noted above.

ztangent replied:
On second thought, maybe the partial forward trajectory of complete states should be stored as `f_trajectory`, and the backward trajectory of partial states should be stored as `b_trajectory`. Storing this information would help us check the correctness of the frontier crossing code more easily using a unit test.

ztangent commented on test/planners.jl:
```diff
@@ -158,6 +158,15 @@ spec = SymbolicPlanners.BackwardSearchGoal(bw_spec, bw_state)
+  spec = SymbolicPlanners.BackwardSearchGoal(bw_spec, bw_state)
+  @test is_goal(spec, blocksworld, sol.trajectory[1])
+  @test collect(sol) == @pddl("(pick-up a)", "(stack a b)",
+                              "(pick-up c)", "(stack c a)")
```
I suspect that one reason why this test is currently passing is because the BiDirectionalPlanner will eventually succeed as long as either the forward search succeeds or the backward search succeeds, even if the frontier crossing code isn't working quite right.

I think it would be good to design a more rigorous test case that actually checks if the frontier crossing code is working -- maybe this could be done by inspecting the search trees in the returned solution and seeing whether they cross?

pevnak commented:
Hi Xuan,

Thanks a lot. I will try to address those, not sure which time-frame, but will do my best. I have to confess that I do not understood 100 % of all details of the library.

pevnak commented:
Hi Xuan,

I have implemented the requested changes. It is not going to be super fast, as I have used your second recommended approach that relies on `issubset` but I have decided to make it right first and then improve the speed.

ztangent commented on Project.toml:
```diff
@@ -4,6 +4,7 @@ authors = ["Xuan <tanqazx@gmail.com>"]
 version = "0.1.9"
 
 [deps]
+BinaryDecisionDiagrams = "cd45f48d-2bff-4983-a793-756749635bd1"
```
Since we're not using BDDs for now, maybe this can be removed?

pevnak replied:
Yes, this is leftover of my experiments. I would love to look at them, but my idea was to start simply first and then to improve. To benchmark it properly. But then, I got a bit stuck how to deal with fluents. Do you know, if there is some implementation in other language, where I can get inspiration?

ztangent replied:
Unfortunately I don't really know of existing implementations!

I think using BDDs to represent partial states is a bigger project that would require modifying PDDL.jl itself. Right now the representation for partial states is actually quite limited, in that it can't handle numeric fluents at all. So the `regress` and `relevant` functions don't work correctly for domains with fluents, which means that `BackwardPlanner` also does not work for domains with fluents. Addressing this would require some form of symbolic execution, such that numeric fluents can be assigned placeholder values. I would love to support something like that in the future, but it's not a priority unfortunately!

ztangent commented on test/planners.jl:
```diff
@@ -158,6 +158,15 @@ spec = SymbolicPlanners.BackwardSearchGoal(bw_spec, bw_state)
 
+@testset "BiDirectional Planner" begin
+  planner = BidirectionalPlanner(HAdd(), HAddR())
+  sol = planner(blocksworld, bw_state, bw_spec)
```
Can we also add a test here to see if the original specification is satisfied?

That is:
```julia
@test is_goal(bw_spec, blocksworld, sol.trajectory[end])
```

pevnak replied:
Certainly, it is added.

ztangent reviewed (changes requested):
Left two minor comments, but apart from that, this looks great to me!

pevnak commented:
I have included the suggested changes.

I have noticed a new api growing around `PathSearchSolution`. I have hijacked that a bit to save a typing. Can you take a look on that?
 I have originally thought about introducing `AbstractSearchSolution`, but I did not know, if it would be something you like and therefore opted just for
`const SearchSolutions = Union{PathSearchSolution, BiPathSearchSolution}`
I have also implemented `Base.copy` for the `BiPathSearchSolution` but due to large number of fields, I made my life easier by using `fieldnames`. Not sure if it is not against your style, I would be very happy to change that back at your wish.

Finally, added a summary print of solutions, since it was very verbose. Again, happy to remove it.

Thanks for patience.

Tomas

ztangent commented on src/planners/path_search.jl:
```diff
@@ -56,14 +56,53 @@ function Base.copy(sol::PathSearchSolution)
+    return BiPathSearchSolution(fields...)
+end
+
+const SearchSolutions = Union{PathSearchSolution, BiPathSearchSolution}
```
Let's replace with this with an abstract type called `AbstractPathSearchSolution` (I don't want to call it `AbstractSearchSolution` because there are other kinds of search algorithms that don't necessarily fit this mold.)

ztangent replied:
Oh, and let's make sure that `AbstractPathSearchSolution` is a subtype of `OrderedSolution`.

pevnak replied:
Done

ztangent commented on src/planners/path_search.jl:
```diff
@@ -56,14 +56,53 @@ function Base.copy(sol::PathSearchSolution)
+
+function Base.show(io::IO, sol::SearchSolutions)
+    pl = sol.status == :success ? length(sol.plan) : "-"
+    println(io, "BiPathSearchSolution: (", sol.status, ", ", pl, ", ", sol.expanded, ")")
```
This is nice to have! But can we:

- Change it to display the name of the type, rather than hardcoding `"BiPathSearchSolution"`? I think you can just do this with `typeof(sol)`.

- Display the plan even if the status is not `:success`? There are cases where the partial plan is helpful, even if you reach the `:max_time` or `:max_nodes` termination criterion. The only case where we shouldn't display the plan is when `sol.status == :failure`.

ztangent replied:
Oh, also, it'd be good to display not just the length of the plan but the plan itself, and maybe indicate clearly what the numbers etc refer to.

Perhaps let's just remove this for now, I want to think a bit more about the best format for displaying a summary of the information.

pevnak replied:
Removed.

ztangent reviewed (changes requested):
Thanks for these changes and suggestions! I'm basically fine with everything, just a few minor things I think we could change.

ztangent commented:
Looks great! Thanks again for contributing this :)

ztangent commented:
FYI, I belatedly realized that there were still some issues with how the plans were being reconstructed when a frontier crossing is detected (I think somehow the test cases must have missed them). They're fixed now 8622d7e.
